[Fingerprint_grow] Implements Sleep Mode feature

alexborro opened the pull request:
# What does this implement/fix?

This feature enable the use of one GPIO pin to toggle the sensor power off when idling for a period of time.
It will be turned on again when there is a request for a communication to the sensor.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** Implements https://github.com/esphome/feature-requests/issues/1672

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3558

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
uart:
  rx_pin: GPIO16
  tx_pin: GPIO17
  baud_rate: 57600
    
fingerprint_grow:
  sensing_pin: GPIO4
  sensor_power_pin:
      number: GPIO18
      inverted: true
  idle_period_to_sleep: 5s
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

alexborro committed:
Implement Sleep Mode feature
  This feature enable the use of one GPIO pin to toggle
  the sensor power off when idling for a period of time.
  It will be turned on again when there is a request for
  a communication to the sensor

OnFreund commented on esphome/components/fingerprint_grow/fingerprint_grow.cpp:
```diff
@@ -344,20 +371,20 @@ uint8_t FingerprintGrowComponent::send_command_() {
   this->write((uint8_t) (sum & 0xFF));
 
-  this->data_.clear();
+  p_data_buffer->clear();
```
I'm not sure clearing out a client supplied buffer is the right approach...

alexborro replied:
Well, I kind of agree with you, but the original code was already using the member `data_` to receive and return data. 

And the `transfer_()`  argument is not `const` so the caller cannot expect to pass a pointer and have no changes on it (otherwise the argument should have the `const` keyword). This whole function can be refactored to fix this original bug but it would require a major refactoring. 

I have a plan to change this whole communication to an asynchronous process, so it no longer keeps the component executing for more than one second without returning.

OnFreund replied:
Yeah, but the original code operated under the assumption that there was only one message at a time, so reusing one buffer was more efficient. It's a bit different now.
In any case, if you plan on getting this communication overhaul done soon I wouldn't sweat it now.

alexborro replied:
Yeap, that is the idea. I didn't want to refactor the whole component communication procedure because it is huge and should be done as a separate task - that is my plan. This is just a new feature some fellows are asking for - including me.

alexborro replied:
Hi @OnFreund 
If you agree on this idea (refactoring this send command function as a next step), could you please approve this PR? Otherwise It is hard for me keeping rebasing my development branches on top of this one (including that ongoing refactoring).
Thanks.

OnFreund replied:
This looks good to me, but it's @loongyh who really knows this code, not me...

loongyh replied:
We need a maintainer who has write access to approve. @jesserockz can you proceed to review and merge?

loongyh reviewed (approved):
Hmm interesting, I have always wondered why the touch pin needs to be seperately powered with `3.3VT` and what's the use case with it. Although I'm currently unable to test it myself, the code LGTM.

 Thanks for your contribution!

alexborro commented:
Hi @jesserockz , could you please check if this PR is good to be merged ? It is lingering here for a few weeks. And it also fixes some reported issues (https://github.com/esphome/issues/issues/5502)

alexborro committed:
Fixes issue #5507
Check if fingerprint sensors are initialized (requested in yaml file) before reporting them.

alexborro commented:
Hi @jesserockz , could you please check if this PR is good enough for the next release ? It is lingering here for weeks. It also fixes some issues so people don't need to keep using external component (issues mentioned in comments).
Cheers.
Alex.
